builder/dockerfile/evaluator.go

```diff
@@ -56,10 +58,32 @@ var allowWordExpansion = map[string]bool{
 	command.Expose: true,
 }
 
-var evaluateTable map[string]func(*Builder, []string, map[string]bool, string) error
+type dispatchRequest struct {
+	builder    *Builder // TODO: replace this with a smaller interface
```

Comment:
`parser.Result` is basically just ast after tokenizing. To actually put together a build definition this is parsed again in dispatchers and then executed(eg https://github.com/moby/moby/blob/8cb9d7ec2e56ca360da55d66bc77e799faca1dd0/builder/dockerfile/dispatchers.go#L185-L197). Instead this should be split up so that static analysis can be performed on the definition and no extra parsing is needed when the actual commands run.

I'm not sure about the intent to limit the interfaces for internal functions. The only public interface to the builder should be a single `Build(ctx, ...) error` function(ignoring uses of dry-run atm).